From b9def9d57059b3aaeba4d5892d05167b8b556ae2 Mon Sep 17 00:00:00 2001
From: Simon Dinkin <simonx.dinkin@intel.com>
Date: Tue, 30 Aug 2016 11:05:10 +0300
Subject: [PATCH 65/69] anqp: add support for neighbor report anqp element

if rrm_neighbor_report is enabled, then in response to
anqp capability list request, the AP will respond with
anqp capability list response that include support for anqp neighbor report.

if there is neighbors in the AP neighbors database, then in response to
anqp neighbor report request, the AP will respond with
anqp neighbor report response that include all the existing neighbors.
if there is no neighbors, the ANQP neighbor report element will be
included with no elements (length 0)
---
 src/ap/gas_serv.c | 30 +++++++++++++++++++++++++++++-
 src/ap/gas_serv.h |  2 ++
 2 files changed, 31 insertions(+), 1 deletion(-)

diff --git a/src/ap/gas_serv.c b/src/ap/gas_serv.c
index 6ce178d..b468856 100644
--- a/src/ap/gas_serv.c
+++ b/src/ap/gas_serv.c
@@ -257,7 +257,8 @@ static void anqp_add_capab_list(struct hostapd_data *hapd,
 		wpabuf_put_le16(buf, ANQP_EMERGENCY_ALERT_URI);
 	if (get_anqp_elem(hapd, ANQP_EMERGENCY_NAI))
 		wpabuf_put_le16(buf, ANQP_EMERGENCY_NAI);
-	if (get_anqp_elem(hapd, ANQP_NEIGHBOR_REPORT))
+	if ((hapd->conf->radio_measurements[0] & WLAN_RRM_CAPS_NEIGHBOR_REPORT) ||
+		get_anqp_elem(hapd, ANQP_NEIGHBOR_REPORT))
 		wpabuf_put_le16(buf, ANQP_NEIGHBOR_REPORT);
 	for (id = 273; id < 277; id++) {
 		if (get_anqp_elem(hapd, id))
@@ -547,6 +548,26 @@ static void anqp_add_domain_name(struct hostapd_data *hapd, struct wpabuf *buf)
 	}
 }
 
+static void anqp_add_neighbor_report(struct hostapd_data *hapd,
+					   struct wpabuf *buf)
+{
+	struct hostapd_neighbor_entry *nr;
+
+	if (anqp_add_override(hapd, buf, ANQP_NEIGHBOR_REPORT))
+		return;
+
+	if (dl_list_empty(&hapd->nr_db)) {
+		wpabuf_put_le16(buf, ANQP_NEIGHBOR_REPORT);
+		wpabuf_put_le16(buf, 0);
+	} else {
+		dl_list_for_each(nr, &hapd->nr_db, struct hostapd_neighbor_entry,
+				list) {
+			wpabuf_put_le16(buf, ANQP_NEIGHBOR_REPORT);
+			wpabuf_put_le16(buf, wpabuf_len(nr->nr));
+			wpabuf_put_buf(buf, nr->nr);
+		}
+	}
+}
 
 #ifdef CONFIG_HS20
 
@@ -859,6 +880,8 @@ gas_serv_build_gas_resp_payload(struct hostapd_data *hapd,
 		anqp_add_elem(hapd, buf, ANQP_TDLS_CAPABILITY);
 	if (request & ANQP_REQ_EMERGENCY_NAI)
 		anqp_add_elem(hapd, buf, ANQP_EMERGENCY_NAI);
+	if (request & ANQP_REQ_NEIGHBOR_REPORT)
+		anqp_add_neighbor_report(hapd, buf);
 
 	for (i = 0; i < num_extra_req; i++)
 		anqp_add_elem(hapd, buf, extra_req[i]);
@@ -983,6 +1006,11 @@ static void rx_anqp_query_list_id(struct hostapd_data *hapd, u16 info_id,
 			     "Emergency NAI",
 			     get_anqp_elem(hapd, info_id) != NULL, qi);
 		break;
+	case ANQP_NEIGHBOR_REPORT:
+		set_anqp_req(ANQP_REQ_NEIGHBOR_REPORT,
+			     "Neighbor Report",
+				 !dl_list_empty(&hapd->nr_db), qi);
+		break;
 	default:
 		if (!get_anqp_elem(hapd, info_id)) {
 			wpa_printf(MSG_DEBUG, "ANQP: Unsupported Info Id %u",
diff --git a/src/ap/gas_serv.h b/src/ap/gas_serv.h
index 9051e4f..4ec852b 100644
--- a/src/ap/gas_serv.h
+++ b/src/ap/gas_serv.h
@@ -40,6 +40,8 @@
 	(1 << (ANQP_TDLS_CAPABILITY - ANQP_QUERY_LIST))
 #define ANQP_REQ_EMERGENCY_NAI \
 	(1 << (ANQP_EMERGENCY_NAI - ANQP_QUERY_LIST))
+#define ANQP_REQ_NEIGHBOR_REPORT \
+	(1 << (ANQP_NEIGHBOR_REPORT - ANQP_QUERY_LIST))
 /*
  * First 16 Hotspot 2.0 vendor specific ANQP-elements can be included in the
  * optimized bitmap.
-- 
1.8.3.4

